File: tools/debug_v1.py
"""Merveil Debug V1 shared diagnosis helpers.
Safe by default: never mutates projects, auth, databases, or deployments.
"""
from __future__ import annotations
import json, re
from pathlib import PurePosixPath

SECRET_FILE = re.compile(r"(^|/)(\.env(?:\..*)?|.*\.pem|.*\.key|.*\.p12|.*\.pfx|credentials?\.json)$", re.I)
TEXT_EXT = {'.js','.jsx','.ts','.tsx','.mjs','.cjs','.json','.css','.html','.md','.sql','.yml','.yaml','.py','.java','.vue','.svelte'}


def safe_path(path: str) -> bool:
    p = str(path or '').replace('\\','/').lstrip('/')
    return '..' not in PurePosixPath(p).parts and not SECRET_FILE.search(p)


def scan_files(files: list[dict]) -> dict:
    issues=[]; paths=[]; total=0
    for item in files[:500]:
        path=str(item.get('path','')).replace('\\','/')
        content=str(item.get('content',''))
        if not safe_path(path) or len(content)>500_000 or PurePosixPath(path).suffix.lower() not in TEXT_EXT:
            continue
        paths.append(path); total += len(content)
        for n,line in enumerate(content.splitlines(),1):
            if re.search(r'(api[_-]?key|secret|access[_-]?token|refresh[_-]?token)\s*[:=]\s*["\'][^"\']{8,}', line, re.I):
                issues.append({'severity':'error','code':'SECRET_LIKE_VALUE','message':'Possible hard-coded secret detected. Remove it and use server-side environment configuration.','file':path,'line':n})
            if 'TODO' in line or 'FIXME' in line:
                issues.append({'severity':'warn','code':'TODO','message':'Unfinished work marker found.','file':path,'line':n})
            if re.search(r'console\.(log|debug)\(',line) and not path.endswith(('.test.js','.test.ts')):
                issues.append({'severity':'info','code':'CONSOLE_LOG','message':'Console logging should be reviewed before production.','file':path,'line':n})
    names=set(paths)
    if not any(p in names for p in ('package.json','pyproject.toml','requirements.txt')):
        issues.append({'severity':'error','code':'NO_MANIFEST','message':'No recognized project dependency manifest was found.'})
    if any(p.startswith('src/') for p in names) and not any(p in names for p in ('src/main.tsx','src/main.jsx','src/main.ts','src/index.tsx','src/index.jsx')):
        issues.append({'severity':'warn','code':'ENTRYPOINT_REVIEW','message':'A src directory exists but no common application entry point was detected.'})
    errors=sum(i['severity']=='error' for i in issues); warns=sum(i['severity']=='warn' for i in issues)
    score=max(0,100-errors*18-warns*6-sum(i['severity']=='info' for i in issues))
    return {'score':score,'status':'Healthy' if score>=90 else 'Review needed' if score>=70 else 'Problems detected','filesScanned':len(paths),'bytesScanned':total,'counts':{'error':errors,'warn':warns,'info':sum(i['severity']=='info' for i in issues)},'issues':issues[:100]}


def redact_payload(value):
    """Return a safe diagnostic representation without credential values."""
    if isinstance(value, dict):
        return {k: '[REDACTED]' if re.search(r'(password|token|secret|authorization|cookie|api[_-]?key|private[_-]?key)', str(k), re.I) else redact_payload(v) for k,v in value.items()}
    if isinstance(value, list): return [redact_payload(v) for v in value[:100]]
    return value

File: server/debug-v1.js
const SKIP = /(^|\/)(node_modules|dist|build|\.git|\.next|coverage)(\/|$)/i;
const SECRET_FILE = /(^|\/)(\.env(?:\..*)?|.*\.(pem|key|p12|pfx)|credentials?\.json)$/i;
const TEXT = /\.(js|jsx|ts|tsx|json|css|html|md|sql|mjs|cjs|vue|py|java|yml|yaml)$/i;
const MAX_FILES = 500;
const MAX_BYTES = 500_000;

export function safePath(path) {
  const p = String(path || '').replace(/\\/g, '/').replace(/^\/+/, '');
  return !!p && !p.split('/').includes('..') && !SECRET_FILE.test(p) && !SKIP.test(p);
}

export function normalizeFiles(files) {
  return (Array.isArray(files) ? files : []).slice(0, MAX_FILES).map(f => ({
    path: String(f?.path || '').replace(/\\/g, '/').replace(/^\/+/, ''),
    content: String(f?.content || '')
  })).filter(f => safePath(f.path) && f.content.length <= MAX_BYTES && TEXT.test(f.path));
}

function issue(severity, code, message, file = null, line = null, fix = null) {
  return { severity, code, message, file, line, fix };
}

export function diagnose(files) {
  const textFiles = normalizeFiles(files);
  const names = new Set(textFiles.map(f => f.path));
  const issues = [];
  if (!names.has('package.json')) issues.push(issue('error', 'NO_MANIFEST', 'No package.json was found in the supplied project.', null, null, 'Add or identify the project package manifest.'));
  if (!names.has('src/App.jsx') && !names.has('src/App.tsx') && !names.has('src/main.jsx') && !names.has('src/main.tsx') && !names.has('index.html')) issues.push(issue('warn', 'ENTRY_NOT_FOUND', 'No common Vite/React entry file was detected.', null, null, 'Confirm the application entry point.'));

  const envRefs = new Set();
  for (const f of textFiles) {
    f.content.split(/\r?\n/).forEach((line, idx) => {
      const n = idx + 1;
      if (/(api[_-]?key|secret|access[_-]?token|refresh[_-]?token|password)\s*[:=]\s*["'][^"']{8,}/i.test(line)) issues.push(issue('error', 'SECRET_LIKE_VALUE', 'Possible hard-coded secret detected. Remove it and use server-side environment configuration.', f.path, n, 'Move the value to server-side environment configuration.'));
      if (/import\.meta\.env\.([A-Z0-9_]+)/.test(line)) envRefs.add(line.match(/import\.meta\.env\.([A-Z0-9_]+)/)[1]);
      if (/process\.env\.([A-Z0-9_]+)/.test(line)) envRefs.add(line.match(/process\.env\.([A-Z0-9_]+)/)[1]);
      if (/TODO|FIXME|XXX/.test(line)) issues.push(issue('warn', 'TODO_MARKER', 'Unfinished marker found.', f.path, n, 'Review before shipping.'));
      if (/console\.(log|debug|error)\(/.test(line) && !/\.test\.[jt]sx?$/.test(f.path)) issues.push(issue('info', 'CONSOLE_LOG', 'Production console logging should be reviewed.', f.path, n, 'Keep only intentional structured logging.'));
      if (/catch\s*\([^)]*\)\s*\{\s*\}/.test(line)) issues.push(issue('warn', 'EMPTY_CATCH', 'An empty catch block can hide a runtime failure.', f.path, n, 'Handle or report the error explicitly.'));
    });
  }

  const pkg = textFiles.find(f => f.path === 'package.json');
  if (pkg) {
    try {
      const manifest = JSON.parse(pkg.content);
      const deps = { ...(manifest.dependencies || {}), ...(manifest.devDependencies || {}) };
      for (const f of textFiles) for (const line of f.content.split(/\r?\n/)) {
        const m = line.match(/(?:import|require)\s*(?:\([^)]*\)|[^;]*?from)?\s*["']([^"'./][^"']*)["']/);
        if (!m) continue;
        const root = m[1].startsWith('@') ? m[1].split('/').slice(0, 2).join('/') : m[1].split('/')[0];
        if (!['react', 'react-dom', 'vite', 'node:'].includes(root) && !deps[root]) issues.push(issue('error', 'MISSING_DEPENDENCY', `Imported package "${root}" is not declared in package.json.`, f.path, null, `Add ${root} to dependencies or remove the import.`));
      }
    } catch { issues.push(issue('error', 'INVALID_PACKAGE_JSON', 'package.json could not be parsed as JSON.', 'package.json', null, 'Fix JSON syntax before building.')); }
  }
  if (envRefs.size) issues.push(issue('info', 'ENV_REFERENCES', `${envRefs.size} environment variable reference(s) detected. Values were not collected.`));

  const counts = { error: 0, warn: 0, info: 0 };
  for (const x of issues) counts[x.severity] = (counts[x.severity] || 0) + 1;
  const score = Math.max(0, Math.min(100, 100 - counts.error * 18 - counts.warn * 6 - counts.info));
  return {
    name: 'merveil-debug-v1', version: 1,
    score,
    status: score >= 90 ? 'Healthy' : score >= 70 ? 'Review needed' : 'Problems detected',
    filesScanned: textFiles.length,
    counts,
    issues: issues.slice(0, 200)
  };
}

export function propose(diagnosis) {
  const issues = Array.isArray(diagnosis?.issues) ? diagnosis.issues : [];
  return {
    pipeline: 'propose',
    readOnly: true,
    proposals: issues.filter(x => x.severity !== 'info').slice(0, 100).map((x, i) => ({
      id: `DBG-${String(i + 1).padStart(3, '0')}`,
      code: x.code,
      severity: x.severity,
      file: x.file,
      line: x.line,
      problem: x.message,
      action: x.fix || 'Review and repair the issue, then run verification again.',
      verification: `Re-run Debug V1 and confirm ${x.code} is no longer reported.`
    }))
  };
}

export function verify(before, after) {
  const beforeIssues = Array.isArray(before?.issues) ? before.issues : [];
  const afterIssues = Array.isArray(after?.issues) ? after.issues : [];
  const afterCodes = new Set(afterIssues.map(x => `${x.code}|${x.file || ''}|${x.line || ''}`));
  const results = beforeIssues.filter(x => x.severity !== 'info').map(x => {
    const key = `${x.code}|${x.file || ''}|${x.line || ''}`;
    return { code: x.code, file: x.file, line: x.line, status: afterCodes.has(key) ? 'still_present' : 'verified_fixed' };
  });
  return {
    pipeline: 'verify',
    verified: results.filter(x => x.status === 'verified_fixed').length,
    remaining: results.filter(x => x.status === 'still_present').length,
    status: results.some(x => x.status === 'still_present') ? 'Review needed' : 'Verified',
    results,
    beforeScore: before?.score ?? null,
    afterScore: after?.score ?? null
  };
}
